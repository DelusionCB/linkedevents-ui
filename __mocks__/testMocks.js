export function mockCurrentTime(mockTime) {
    const RealDate = Date

    class MockDate extends RealDate {
        constructor(values) {
            super()
            if (values) {
                return new RealDate(values)
            } else {
                return new RealDate(mockTime)
            }
        }

        static now() {
            return (new RealDate(mockTime)).getTime()
        }
    }
    global.Date = MockDate
}

export const resetMockDate = () => {
    global.Date = window.Date
}

/**
 *
 * @param initialState
 * @param props
 * @returns {*}
 */
export function getProps(initialState, props){
    return {...initialState, ...props}
}
