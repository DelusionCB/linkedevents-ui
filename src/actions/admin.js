import constants from '../constants'
import client from '../api/client';
import {setFlashMsg} from './app';
import {sendDataComplete,setLoading} from './editor';
import {scrollToTop} from '../utils/helpers';


export async function makeOrgUserRequest(id, pageSize, pageNumber = null, filter) {
    const params = {
        organization: id,
        page_size: pageSize,
        page: pageNumber,
        full_name: filter,
    }
    try {
        const response = await client.get('user', params);

        // Append the page number to the JSON array so that it can be easily used in the pagination
        if (pageNumber !== null) {
            response.data.meta.currentPage = pageNumber;
        }
        return response.data
    } catch (e) {
        throw Error(e)
    }
}

// Fetch AdminOptions
export const fetchAdminOptions = () => async (dispatch) => {
    try {
        const responseOne = await client.get('organization_class')
        const options = responseOne.data.data
        dispatch(receiveAdminOptions(options))
    } catch (e) {
        throw Error(e)
    }
}

// Receive AdminOptions
export const receiveAdminOptions = (data) => {
    return {
        type: constants.ADMIN.RECEIVE_ADMINDATA,
        data,
    }
}

export const formatValues = (data, updating) => {

    if (data.sub_organizations && data.sub_organizations.length !== 0) {
        data.sub_organizations = data.sub_organizations.reduce((acc, curr) => {
            acc.push(curr['@id'])
            return acc
        }, [])
    }
    if (!data.parent_organization.length) {
        data.parent_organization = undefined
    }
    if (data.founding_date === '') {
        data.founding_date = undefined
    }
    if (!updating) {
        data.origin_id = ''
    }
    data.affiliated_organizations = []
    data.replaced_by = ''
    return data
}


export const executeSendRequestOrg = (
    orgValues,
    updatingOrganization,
) => async (dispatch) => {

    let preparedOrgValues

    if (!orgValues) {
        return
    } else {
        preparedOrgValues = formatValues(orgValues, updatingOrganization)
    }


    try {
        const url = updatingOrganization
            ? `organization/${orgValues.id}/`
            : 'organization/'

        const response = updatingOrganization
            ? await client.put(url, preparedOrgValues)
            : await client.post(url, preparedOrgValues)


        let data = response.data
        // validation errors
        if (response.status === 400) {
            data.apiErrorMsg = 'validation-error'
            data.response = response
        }

        // auth errors
        if (response.status === 401 || response.status === 403) {
            data.apiErrorMsg = 'authorization-required'
            data.response = response
        }
        if (response.status === 201 || response.status === 200) {
            data.apisuccessMsg = updatingOrganization ? 'admin-org-updated' : 'admin-org-created'
        }

        dispatch(sendOrgComplete(data))
    } catch (error) {
        dispatch(setFlashMsg('server-error', 'error', error))
        dispatch(setLoading(false))
        new Error(error)
    }
}

export const executeSendRequestUser = (
    userValues,
) => async (dispatch) => {

    let preparedUserValues

    if (!userValues) {
        return
    } else {
        preparedUserValues = userValues
    }


    try {
        const url = `user/${userValues.username}/update_organization_levels`

        preparedUserValues.username = undefined
        const response = await client.put(url, preparedUserValues)


        let data = response.data

        // validation errors
        if (response.status === 400) {
            data.apiErrorMsg = 'validation-error'
            data.response = response
        }

        // auth errors
        if (response.status === 401 || response.status === 403) {
            data.apiErrorMsg = 'authorization-required'
            data.response = response
        }
        if (response.status === 200) {
            data.apisuccessMsg =  'admin-user-updated'
        }
        dispatch(sendOrgComplete(data))
    } catch (error) {
        dispatch(setFlashMsg('server-error', 'error', error))
        dispatch(setLoading(false))
        new Error(error)
    }
}

export const sendOrgComplete = (data) => (dispatch) => {
    const error = data.apiErrorMsg;
    const flashMsg = {
        msg: error || data.apisuccessMsg,
        style: error ? 'error' : 'message',
        data: data,
    }
    dispatch(setFlashMsg(flashMsg.msg, flashMsg.style, flashMsg.data))
    if (!error) {
        scrollToTop()
    }
}
