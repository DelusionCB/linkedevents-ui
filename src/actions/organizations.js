import constants from '../constants.js'
import client from '../api/client'

const {FETCHING_ORGANIZATIONSDATA, RECEIVE_ORGANIZATIONSDATA, RECEIVE_ORGANIZATIONS_ERROR} = constants

// Handled by the organizations reducer
export function receiveOrganizationsData(data) {
    return {
        type: RECEIVE_ORGANIZATIONSDATA,
        payload: data,
    }
}

export function receiveOrganizationsFail() {
    return {
        type: constants.RECEIVE_ORGANIZATIONS_ERROR,
        payload: [],
    }
}
/**
 * Set organizations.isOrganizationsFetching boolean
 * @param {boolean} data
 * @returns {{payload: boolean, type: string}}
 */
export function loadingOrganizationsData(data) {
    return {
        type: FETCHING_ORGANIZATIONSDATA,
        payload: data,
    }
}

// Handles getting all organizations data from api
export const fetchOrganizations = () => async (dispatch) => {
    let organizationsResponse;
    try {
        dispatch(loadingOrganizationsData(true))
        // fetch organizations from endpoint
        organizationsResponse = await client.get('organization')
        const organizationsData = organizationsResponse.data
        dispatch(receiveOrganizationsData(organizationsData))
    } catch (error) {
        dispatch(loadingOrganizationsData(false))
        dispatch(receiveOrganizationsFail(organizationsResponse))
        new Error(error);
    }
}
