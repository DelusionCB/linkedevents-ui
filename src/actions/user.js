import constants from '../constants.js'
import {set, get} from 'lodash'
import {setEditorAuthFlashMsg} from './editor'
import client from '../api/client'
import {getAdminOrganizations, getRegularOrganizations, getPublicOrganizations} from '../utils/user'
import store from '../store.js'

const {RECEIVE_USERDATA, CLEAR_USERDATA, USER_TYPE, FETCHING_USERDATA, SET_ACTIVE_ORGANIZATION} = constants

// Handled by the user reducer
export function receiveUserData(data) {
    return {
        type: RECEIVE_USERDATA,
        payload: data,
    }
}

// Handled by the user reducer
export function clearUserData() {
    return {
        type: CLEAR_USERDATA,
    }
}

/**
 * Set user.isFetching boolean
 * @param {boolean} data
 * @returns {{payload: boolean, type: string}}
 */
export function loadingUser(data) {
    return {
        type: FETCHING_USERDATA,
        payload: data,
    }
}

export function setActiveOrganization (org) {
    return {
        type: SET_ACTIVE_ORGANIZATION,
        payload: org,
    }
}
const getUserType = (permissions) => {
    if (permissions.includes(USER_TYPE.SUPERADMIN)) {
        return USER_TYPE.SUPERADMIN
    }
    if (permissions.includes(USER_TYPE.ADMIN)) {
        return USER_TYPE.ADMIN
    }
    if (permissions.includes(USER_TYPE.REGULAR)) {
        return USER_TYPE.REGULAR
    }
    if (permissions.includes(USER_TYPE.PUBLIC)) {
        return USER_TYPE.PUBLIC
    }
}

// Handles getting user data from backend api with given id.
export const fetchUser = (id) => async (dispatch) => {
    try {
        dispatch(loadingUser(true))
        // try to get user data from user endpoint
        const userResponse = await client.get(`user/${id}`)
        const userData = userResponse.data
        // add correct permissions to user based on user's organizations
        const permissions = []

        const state = store.getState();
        const {user:{activeOrganization}} = state;
        const adminInOrgs = get(userData, 'admin_organizations', [])
        const regularInOrgs = get(userData, 'organization_memberships', [])
        const publicOrgs = get(userData, 'public_memberships', [])
        const activeOrgId = activeOrganization ?? userData.organization;

        if (adminInOrgs.length > 0 && (adminInOrgs).includes(activeOrgId)) {
            permissions.push(USER_TYPE.ADMIN)
        }
        if (regularInOrgs.length > 0 && (regularInOrgs).includes(activeOrgId)) {
            permissions.push(USER_TYPE.REGULAR)
        }
        if (publicOrgs.length > 0 && (publicOrgs).includes(activeOrgId)) {
            permissions.push(USER_TYPE.PUBLIC)
        }
        if (get(userData, 'is_superuser', false)) {
            permissions.push(USER_TYPE.SUPERADMIN)
        }


        // add all desired user data in an object which will be stored into redux store
        const mergedUser = {
            id: get(userData, 'uuid', null),
            displayName: get(userData, 'display_name', null),
            firstName: get(userData, 'first_name', null),
            lastName: get(userData, 'last_name', null),
            username: get(userData, 'username', null),
            email: get(userData, 'email', null),
            organization: get(userData, 'organization', null),
            adminOrganizations: get(userData, 'admin_organizations', null),
            organizationMemberships: get(userData, 'organization_memberships', null),
            publicMemberships: get(userData, 'public_memberships', null),
            permissions,
            userType: getUserType(permissions),
        }

        const adminOrganizations = await Promise.all(getAdminOrganizations(mergedUser))
        const regularOrganizations = await Promise.all(getRegularOrganizations(mergedUser))
        const publicOrganizations = await Promise.all(getPublicOrganizations(mergedUser))
        // store data of all the organizations that the user is admin in
        mergedUser.adminOrganizationData = adminOrganizations
            .reduce((acc, organization) => set(acc, `${organization.data.id}`, organization.data), {})
        // store data of all the organizations where the user is a regular user
        mergedUser.regularOrganizationData = regularOrganizations
            .reduce((acc, organization) => set(acc, `${organization.data.id}`, organization.data), {})
        // store data of all the organizations where the user is public user
        mergedUser.publicOrganizationData = publicOrganizations
            .reduce((acc, organization) => set(acc, `${organization.data.id}`, organization.data), {})
        // get organizations with regular users
        mergedUser.organizationsWithRegularUsers = adminOrganizations
            .filter(organization => get(organization, ['data', 'has_regular_users'], false))
            .map(organization => organization.data.id)
        dispatch(setActiveOrganization(activeOrgId))
        dispatch(receiveUserData(mergedUser))
        dispatch(setEditorAuthFlashMsg())
    } catch (e) {
        dispatch(loadingUser(false))
        throw Error(e)
    }
}
