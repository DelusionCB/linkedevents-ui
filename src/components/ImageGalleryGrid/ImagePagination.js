import React from 'react';
import PropTypes from 'prop-types';

const ImagePagination = (props) => {
    if (props.responseMetadata === undefined) {
        return null;
    }

    const pageAmount = Math.ceil(parseInt(props.responseMetadata.count) / props.pageSize);
    const currentPage = props.responseMetadata.currentPage;

    let classes;
    const pages = [];
    for (let i = 1; i < pageAmount + 1; i++) {
        classes = (props.responseMetadata.currentPage !== undefined && currentPage === i) ? 'page-item active' : 'page-item';
        const additionalProps = currentPage === i ? {'aria-current':'page'} : {};
        pages.push(
            <li className={classes} {...additionalProps} key={i}>
                <a className='page-link' aria-label={props.intl.formatMessage({id: 'image-page'}, {n: i})} href='#' onClick={(e) => props.clickedPage(i,e)}>{i}</a>
            </li>);
    }
    return <nav role='navigation' aria-label={props.intl.formatMessage({id: `image-pagination`})}>
        <ul className='pagination'>
            {pages}
        </ul>
    </nav>
};

ImagePagination.propTypes = {
    pageSize: PropTypes.number,
    responseMetadata: PropTypes.object,
    clickedPage: PropTypes.func,
    intl: PropTypes.object,
};

export default ImagePagination
