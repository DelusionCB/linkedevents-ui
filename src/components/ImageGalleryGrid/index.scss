.search-images {
    align-items: center;
    display: flex;
    input {
        max-width: 200px;
    }
    button {
        margin-left: 10px;
    }
    .inputIcon {
        border: 2px;
        background-color: var(--main-color);
        align-items: center;
        .glyphicon {
            font-size: 1.375rem;
            top: 0px;
            height: 38px;
            width: 38px;
            background-color: var(--main-color);
            background-repeat: round;
            margin: 0px !important;
            color: white;
            padding: 8px 8px 8px 8px !important;
        }
    }
}
.search-results {
    border-bottom: 2px solid black;
    margin-top: 20px;
}
.image-row {
    overflow-y: scroll;
    height: 500px;
    margin-bottom: 10px;
}
.imageShort {
    overflow: auto;
    height: auto;
}
