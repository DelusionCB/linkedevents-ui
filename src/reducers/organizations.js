import constants from '../constants'
import updater from 'immutability-helper';

const initialState = {
    isOrganizationsFetching: false,
    data: [],
}

function update(state = initialState, action) {
    switch(action.type) {
        // Successfully received organizations data.
        case constants.RECEIVE_ORGANIZATIONSDATA:
            if(action.payload && !!action.payload.data) {
                return updater(state, {
                    isOrganizationsFetching:{$set: false},
                    data: {$set: action.payload.data},
                })
            }
            return state
        // Set boolean displaying if organizationdata is currently being fetched.
        case constants.FETCHING_ORGANIZATIONSDATA:
            return updater(state, {
                isOrganizationsFetching: {$set: action.payload},
            })
        case constants.RECEIVE_ORGANIZATIONS_ERROR:
            return updater(state, {
                isOrganizationsFetching: {$set: false},
                data: {$set: initialState.data},
            })
        default:
            return state
    }
}

export default update
