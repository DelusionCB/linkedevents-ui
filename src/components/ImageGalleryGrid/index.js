import './index.scss';

import React from 'react';
import PropTypes from 'prop-types'
import ImagePagination from './ImagePagination';
import {get as getIfExists} from 'lodash'
import {connect} from 'react-redux'
import {fetchUserImages as fetchUserImagesAction} from 'src/actions/userImages'
import ImageThumbnail from '../ImageThumbnail'
import {Button, Input, Label, Form, InputGroup, InputGroupAddon} from 'reactstrap';
import Spinner from 'react-bootstrap/Spinner';
import {FormattedMessage, injectIntl, intlShape} from 'react-intl';
import classNames from 'classnames';
import moment from 'moment'
import {MultiSelect} from '../../components/MultiSelect/MultiSelect'
import {transformOrganizationData} from '../../utils/helpers'
class ImageGalleryGrid extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            searchString: '',
            selectedPublishers: [],
        }
    }

    componentDidMount() {
        this.fetchImages();
    }

    componentDidUpdate(_, prevState) {    
        if (prevState.selectedPublishers !== this.state.selectedPublishers) {
            this.fetchImages();
        }
    }

    /**
     * Fetch images based on conditionals.
     * default fetch uses pageSize & pageNumber as default params.
     *
     * If defaultModal is true, pass 'true' to fetchUserImages, this sets publicImages true and
     * fetches default images which everyone can use.
     *
     * If user is logged in & searchString-state is not empty string, pass 'false' (publicImages), 'true' (filter) & searchString (filterString)
     * into request. This creates filtered fetch & brings items based on the searchString.
     *
     * @param {object} user
     * @param {number} pageSize Size of the items in request
     * @param {number} pageNumber Number of the page as pageSize is limited
     */
    fetchImages = (user = this.props.user, pageSize = this.props.pageSize, pageNumber = 1) => {
        const {fetchUserImages, defaultModal} = this.props;
        const {searchString, selectedPublishers} = this.state;
        let publisher = null;
        if(user && !!selectedPublishers){
            publisher = selectedPublishers.join(',')
        }
        const parameters = [pageSize, pageNumber,publisher]
        if (defaultModal) {
            parameters.push(true)
        } else if (user && searchString !== '') {
            parameters.push(false, true, searchString.trim())
        }
        fetchUserImages(...parameters);
    };

    /**
     * Get the desired page number as a parameter and fetch images for that page
     * @param {number} pageNumber
     * @param e Used for prevent default
     */
    changeImagePage = (pageNumber, e) => {
        e.preventDefault();
        this.fetchImages(this.props.user, 50, pageNumber);
    };

    /**
     * @param {string} e string to set state
     */
    searchOnChange = (e) => {
        this.setState({searchString: e})
    }
    /**
     * preventDefault & fetch images
     */
    submitHandler = () => {
        event.preventDefault();
        this.fetchImages();
    }

    formatDateTime = (dateTime) => {
        return moment(dateTime).format('D.M.YYYY H.mm')
    }

    /**
     * handles multiple organization selection
     * @param options currently selected organizations
     */
    handleOrganizationValueChange = async (options,e) => {
        if(options){
            const selectOrgs = options.map(item => item.value);
            this.setState((state)=>({
                ...state,
                selectedPublishers: selectOrgs,
            }
            ))
        }
    }

    render() {
        // save the id of the selected image of this event (or editor values)
        let selected_id = getIfExists(this.props.editor.values, 'image.id', null);
        // build the classes for the thumbnails
        let images = this.props.images.items.map((img) => {
            let selected = selected_id === img.id
            return (
                <div className='col-md-3 col-xs-12' key={img.id}>
                    <ImageThumbnail
                        locale={this.props.intl.locale}
                        selected={selected}
                        key={img.id}
                        url={img.url}
                        data={img}
                        defaultModal={this.props.defaultModal}
                        close={this.props.close}
                        user={this.props.user}
                        isSharedImage={img.is_shared_within_org}
                        publisher={img.publisher}
                    />
                    {
                        this.props.showImageDetails && (
                            <div className='imageDetails'> 
                                <p><FormattedMessage id='manage-media-publisher' /> {img.publisher_name}</p>
                                <p><FormattedMessage id='manage-media-created' /> {this.formatDateTime(img.created_time)}</p>
                                <p><FormattedMessage id='manage-media-last-modified' /> {this.formatDateTime(img.last_modified_time)}</p>
                            </div>
                        )
                    }
                </div >
            )
        });
       
        // ...and finally check if there is no image for this event to be able to set the class
        const {isFetching, fetchComplete, meta} = this.props.images
        const {defaultModal, user, intl, pageSize, organizations} = this.props
        const pageAmount = fetchComplete ? Math.ceil(parseInt(meta.count) / pageSize) : null;
        const imageCount = fetchComplete ? meta.count : null;
        const formatedOrganizations = !!organizations && (transformOrganizationData(organizations) ?? []);

        return (
            <div className='image-grid container-fluid'>
                {!defaultModal && user &&
                    <div>
                        <div className='search-images'>
                            <Form onSubmit={this.submitHandler}>
                                <Label htmlFor='search-imgs'><FormattedMessage id='search-images'/></Label>
                                <InputGroup>
                                    <InputGroupAddon className='inputIcon' addonType="prepend">
                                        <span aria-hidden className="glyphicon glyphicon-search"/>
                                    </InputGroupAddon>
                                    <Input
                                        id='search-imgs'
                                        placeholder={intl.formatMessage({id: 'search-images-text'})}
                                        type='text'
                                        onChange={(e) => this.searchOnChange(e.target.value)}
                                        value={this.state.searchString}
                                    />
                                
                                    <Button
                                        color='primary'
                                        variant='contained'
                                        type='submit'>
                                        <FormattedMessage id='search-images-text' />
                                    </Button>
                                </InputGroup>
                                {
                                    this.props.showOrganizationFilter && 
                                      <div className='organizationFilter'>
                                          <Label htmlFor='organization-select'>
                                              <FormattedMessage id='manage-media-filter-by-organization' />
                                          </Label>
                                          <MultiSelect 
                                              data={formatedOrganizations} 
                                              placeholder={intl.formatMessage({id: `organization-select-placeholder`})}
                                              handleChange={this.handleOrganizationValueChange}
                                          />
                                      </div>
                                }
                            </Form>
                        </div>
                        <div role='status' className='search-results'>
                            {isFetching && !fetchComplete
                                ? <div/>
                                : <FormattedMessage id='search-image-result' values={{count: imageCount, page: pageAmount}}>{txt => <p role='status'>{txt}</p>}</FormattedMessage>
                            }
                        </div>
                    </div>
                }
                {isFetching && !fetchComplete
                    ? <div role="status" className="search-loading-spinner"><Spinner animation="border">
                        <span className="sr-only">Loading...</span>
                    </Spinner></div>
                    : <div className={classNames ('row', {'image-row': !defaultModal, 'imageShort': images.length < 4})}>
                        {images.length > 0 ? images : <div/>}
                        <div className='clearfix'/>
                    </div>
                }
                {!defaultModal && (
                    <ImagePagination intl={intl} clickedPage={this.changeImagePage} pageSize={pageSize} responseMetadata={this.props.images.meta} />
                )}
            </div>
        )
    }
}

ImageGalleryGrid.defaultProps = {
    pageSize: 50,
}

ImageGalleryGrid.propTypes = {
    images: PropTypes.any,
    user: PropTypes.object,
    editor: PropTypes.object,
    fetchUserImages: PropTypes.func,
    defaultModal: PropTypes.bool,
    action: PropTypes.func,
    close: PropTypes.func,
    pageSize: PropTypes.number,
    intl: intlShape,
    showImageDetails: PropTypes.bool,
    showOrganizationFilter: PropTypes.bool,
    organizations: PropTypes.array,
};

const mapDispatchToProps = (dispatch) => ({
    fetchUserImages: (amount, pageNumber, publicImages, filter, filterString, publisher) => dispatch(fetchUserImagesAction(amount, pageNumber, publicImages, filter, filterString, publisher)),
});

const mapStateToProps = (state, ownProps) => ({
    organizations: state.organizations.data,
});

export {ImageGalleryGrid as UnconnectedImageGalleryGrid}
// TODO: if leave null, react-intl not refresh. Replace this with better React context
export default connect(mapStateToProps, mapDispatchToProps)(injectIntl(ImageGalleryGrid))
