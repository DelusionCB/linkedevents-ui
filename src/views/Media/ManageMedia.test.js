import configureStore from 'redux-mock-store';
import React from 'react';
import thunk from 'redux-thunk';
import {shallow} from 'enzyme';
import {IntlProvider, FormattedMessage} from 'react-intl';
import testReduxIntWrapper from '../../../__mocks__/testReduxIntWrapper';
import ConnectedManageMedia, {ManageMedia} from './ManageMedia';
import {mockCurrentTime, resetMockDate} from '../../../__mocks__/testMocks';
import {mockImages, mockUser} from '../../../__mocks__/mockData';
import fiMessages from 'src/i18n/fi.json';
import mapValues from 'lodash/mapValues';
import renderer from 'react-test-renderer';
import ImageGalleryGrid from '../../components/ImageGalleryGrid/index';
import {fetchOrganizations} from '../../actions/organizations';

const testMessages = mapValues(fiMessages, (value, key) => value);
const intlProvider = new IntlProvider({locale: 'fi', messages: testMessages}, {});
const {intl} = intlProvider.getChildContext();
const mockStore = configureStore([thunk]);

const mockUserSuperAdmin = mockUser;
mockUserSuperAdmin.userType = 'superadmin'

const initialStore = {
    user: mockUserSuperAdmin,
    app: {
        flashMsg: null,
        confirmAction: null,
    },
    auth: {
        isLoadingUser: false,
        user: {},
    },
};

const defaultProps = {
    intl,
    auth: {
        isLoadingUser: false,
        user: null,
    },
    routerPush: jest.fn(),
    user: mockUserSuperAdmin,
    isFetchingUser: false,
    editor: {
        values: {
                
        },
    },
    images: {
        fetchComplete: true,
        isFetching: false,
        meta: {},
        items: mockImages,
    },
    fetchOrganizations: jest.fn(),
    activeOrganization: 'turku:04',
    fetchUserImages: jest.fn(),

};

describe('Media Snapshot', () => {
    let store;

    beforeEach(() => {
        mockCurrentTime('2022-12-07T12:49:39.690Z');
    });

    afterEach(() => {
        resetMockDate();
    });

    it('should render view by default', () => {
        const componentProps = {
            editor: {
                values: {
                        
                },
            },
            user: mockUserSuperAdmin,
            showImageDetails: true,
            fetchOrganizations: jest.fn(),
            ...defaultProps,
        };
        const wrapper = shallow(<ManageMedia {...componentProps} />, {context: {intl}});
        expect(wrapper).toMatchSnapshot();
    });

    it('should render view correctly', () => {
        store = mockStore(initialStore);
        const componentProps = {
            editor: {
                values: {
                        
                },
            },
            user: mockUserSuperAdmin,
            showImageDetails: true,
            fetchOrganizations: jest.fn(),
            ...defaultProps,

        };
        const componentToTest = <ConnectedManageMedia {...componentProps} />;
        const wrapper = renderer.create(testReduxIntWrapper(store, componentToTest));

        expect(wrapper).toMatchSnapshot();
    });
});

describe('Media', () => {
    function getWrapper(props) {
        return shallow(<ManageMedia {...defaultProps} {...props} />, {context: {intl}});
    }

    describe('lifecycle methods', () => {
        describe('routerPush', () => {
            afterEach(() => {
                jest.clearAllMocks();
            })
            test('is called if user is null && !isFetchingUser && !auth.isLoadingUser && !auth.user', () => {
                const wrapper = getWrapper({user: {}});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
                wrapper.setProps({user: null});
                expect(defaultProps.routerPush).toHaveBeenCalled();
                expect(defaultProps.routerPush).toHaveBeenCalledWith('/');
            });
            test('is not called if user is null && !isFetchingUser && !auth.isLoadingUser && auth.user', () => {
                const wrapper = getWrapper({user: {}});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
                wrapper.setProps({user: null, auth:{isLoadingUser: false, user: {}}});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
            });
            test('is not called if user is null && !isFetchingUser && auth.isLoadingUser', () => {
                const wrapper = getWrapper({user: {}});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
                wrapper.setProps({user: null, auth:{isLoadingUser: true}});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
            });
            test('is not called if user is null && isFetchingUser', () => {
                const wrapper = getWrapper({user: {}});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
                wrapper.setProps({user: null, isFetchingUser: true});
                expect(defaultProps.routerPush).not.toHaveBeenCalled();
            });

            describe('componentDidMount', () =>{
                test('fetchOrganizationsData is called in mount', () => {
                    const wrapper = getWrapper()
                    const instance = wrapper.instance();
                    const spy = jest.spyOn(instance, 'fetchOrganizationsData');
                    instance.componentDidMount()
                    expect(spy).toHaveBeenCalled()
                })
    
            })

            describe('componentDidUpdate', () =>{
                test('fetchOrganizationsData is called on update', () => {
                    const wrapper = getWrapper()
                    const instance = wrapper.instance();
                    const prevProps = {...instance.props, user: null}
                    const prevState = {...instance.state}
                    const spy = jest.spyOn(instance, 'fetchOrganizationsData');
                    instance.componentDidUpdate(prevProps, prevState)
                    expect(spy).toHaveBeenCalled()
                })

                test('fetchImages is called on update', () => {
                    const wrapper = getWrapper()
                    const instance = wrapper.instance();
                    const prevProps = {...instance.props, activeOrganization: 'turku:0719'}
                    const prevState = {...instance.state}
                    const spy = jest.spyOn(instance, 'fetchImages');
                    instance.componentDidUpdate(prevProps, prevState)
                    expect(spy).toHaveBeenCalled()
                })
            })
        });
    })
    describe('renders', () => {
        describe('manage media', () => {
            test('render div', () => {
                const wrapper = getWrapper();
                const div = wrapper.find('div');
                expect(div).toHaveLength(1)
                expect(div.prop('className')).toBe('container manageMedia')
            });
            
            test('message', () => {
                const wrapper = getWrapper();
                const message = wrapper.find('#manage-media');
                expect(message).toHaveLength(1)
            });
        });
    });
});
