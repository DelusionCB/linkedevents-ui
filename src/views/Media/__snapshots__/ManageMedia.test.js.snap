// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media Snapshot should render view by default 1`] = `
<div
  className="container manageMedia"
>
  <h1>
    <FormattedMessage
      id="manage-media"
      values={Object {}}
    />
  </h1>
  <Connect(InjectIntl(ImageGalleryGrid))
    editor={
      Object {
        "values": Object {},
      }
    }
    images={
      Object {
        "fetchComplete": true,
        "isFetching": false,
        "items": Array [
          Object {
            "@context": "http://schema.org",
            "@id": "https://api.hel.fi/linkedevents-test/v1/image/47696/",
            "@type": "ImageObject",
            "created_time": "2018-11-16T15:04:44.720111Z",
            "cropping": "",
            "data_source": "helmet",
            "id": 47696,
            "last_modified_time": "2018-11-16T15:04:44.720130Z",
            "license": "cc_by",
            "name": "",
            "photographer_name": null,
            "publisher": "ahjo:u4804001010",
            "url": "https://www.helmet.fi/download/noname/{482BADF8-A86A-4D10-BF30-4FE0DC7A73E7}/69190",
          },
          Object {
            "@context": "http://schema.org",
            "@id": "https://api.hel.fi/linkedevents-test/v1/image/47695/",
            "@type": "ImageObject",
            "created_time": "2018-11-16T15:04:43.074981Z",
            "cropping": "",
            "data_source": "helmet",
            "id": 47695,
            "last_modified_time": "2018-11-16T15:04:43.075001Z",
            "license": "cc_by",
            "name": "",
            "photographer_name": null,
            "publisher": "ahjo:u4804001010",
            "url": "https://www.helmet.fi/download/noname/{BCA5FC49-8479-458A-8C50-ED041623837B}/69214",
          },
          Object {
            "@context": "http://schema.org",
            "@id": "https://api.hel.fi/linkedevents-test/v1/image/47694/",
            "@type": "ImageObject",
            "created_time": "2018-11-16T15:04:42.563744Z",
            "cropping": "",
            "data_source": "helmet",
            "id": 47694,
            "last_modified_time": "2018-11-16T15:04:42.563762Z",
            "license": "cc_by",
            "name": "",
            "photographer_name": null,
            "publisher": "ahjo:u4804001010",
            "url": "https://www.helmet.fi/download/noname/{35F0C5D4-BD73-4D4F-B798-1F3D9C530AAE}/69221",
          },
          Object {
            "@context": "http://schema.org",
            "@id": "https://api.hel.fi/linkedevents-test/v1/image/47693/",
            "@type": "ImageObject",
            "created_time": "2018-11-16T15:04:41.097735Z",
            "cropping": "",
            "data_source": "helmet",
            "id": 47693,
            "last_modified_time": "2018-11-16T15:04:41.097753Z",
            "license": "cc_by",
            "name": "",
            "photographer_name": null,
            "publisher": "ahjo:u4804001010",
            "url": "https://www.helmet.fi/download/noname/{F1E07D11-EA41-4343-AD0E-7413789AF15E}/69199",
          },
        ],
        "meta": Object {},
      }
    }
    showImageDetails={true}
    user={
      Object {
        "_json": Object {
          "date_joined": "2017-11-27T08:07:41.123918Z",
          "department_name": null,
          "email": "foo@foo.com",
          "first_name": "",
          "last_login": "2018-01-19T07:33:42.102250Z",
          "last_name": "",
          "username": "u-ajbd4wwtjii6pgsbyks5pa3yvq",
          "uuid": "0a423e5a-d34a-11e7-9a41-c2a5d78378ac",
        },
        "_raw": "{\\"last_login\\":\\"2018-01-19T07:33:42.102250Z\\",\\"username\\":\\"u-ajbd4wwtjii6pgsbyks5pa3yvq\\",\\"email\\":\\"foo@foo.com\\",\\"date_joined\\":\\"2017-11-27T08:07:41.123918Z\\",\\"first_name\\":\\"\\",\\"last_name\\":\\"\\",\\"uuid\\":\\"0a423e5a-d34a-11e7-9a41-c2a5d78378ac\\",\\"department_name\\":null}",
        "adminOrganizationData": Object {
          "ahjo:02100": Object {
            "id": "ahjoID",
          },
        },
        "emails": Array [
          Object {
            "value": "foo@foo.com",
          },
        ],
        "firstName": "",
        "id": "0a423e5a-d34a-11e7-9a41-c2a5d78378ac",
        "lastName": "",
        "organization": "ahjo:02100",
        "provider": "helsinki",
        "token": "sometoken",
        "userType": "superadmin",
        "username": "u-bjbd4wwtjii6pgsbyks5pa3yvq",
      }
    }
  />
</div>
`;

exports[`Media Snapshot should render view correctly 1`] = `
<div
  className="container manageMedia"
>
  <h1>
    <span>
      Median hallinta
    </span>
  </h1>
</div>
`;
